contracts/ColonyNetworkStaking.sol

```diff
@@ -0,0 +1,306 @@
+  // they failed), or it might be that they timed out
+  mapping (uint256 => uint256) nHashesCompletedChallengeRound;
+  // A flaw with this is that if someone spams lots of nonsense transactions, then 'good' users still have to come along and
+  // explicitly complete the pairings. But if they get the tokens that were staked in order to make the submission, maybe
+  // that's okay...?
+
+  uint256 public nSubmittedHashes = 0;
+  uint256 public nInvalidatedHashes = 0;
+
+  struct Submission {
+    bytes32 hash;
+    uint256 nNodes;
+    uint256 lastResponseTimestamp;
+    uint256 challengeStepCompleted;
+  }
+
+  // Records for which hashes, for which addresses, for which entries have been accepted
+  // Otherwise, people could keep submitting the same entry.
+  mapping (bytes32 => mapping(address => mapping(uint => bool))) submittedEntries;
+
+  event Hash(bytes32 hash);
+
+  function ReputationMiningCycle() public {
+    colonyNetworkAddress = msg.sender;
+    reputationMiningWindowOpenTimestamp = now;
+  }
+
+  function respondToChallenge(uint256 round, uint256 idx) public {
+    // TODO: Check challenge response is valid, relating to current challenge
+    // TODO: Check challenge response relates to this hash
+
+    // Assuming that the challenge response is correct...
+    uint256 opponentIdx = (idx % 2 == 1 ? idx-1 : idx + 1);
+
+    disputeRounds[round][idx].lastResponseTimestamp = now;
+    disputeRounds[round][idx].challengeStepCompleted += 1;
+    // If our opponent responded to this challenge before we did, we should
+    // reset their 'last response' time to now, as they aren't able to respond
+    // to the next challenge before they know what it is!
+    if (disputeRounds[round][idx].challengeStepCompleted == disputeRounds[round][opponentIdx].challengeStepCompleted) {
+      disputeRounds[round][opponentIdx].lastResponseTimestamp = now;
+    }
+  }
+
+  function submitNewHash(bytes32 newHash, uint256 nNodes, uint256 entry) public {
+    //Check the ticket is an eligible one for them to claim
+    require(entry <= IColonyNetwork(colonyNetworkAddress).getStakedBalance(msg.sender) / 10**15);
+    require(entry > 0);
+    if (hasSubmitted[msg.sender].hash != 0x0) {           // If this user has submitted before during this round...
+      require(newHash == hasSubmitted[msg.sender].hash);  // ...require that they are submitting the same hash ...
+      require(nNodes == hasSubmitted[msg.sender].nNodes); // ...require that they are submitting the same number of nodes for that hash ...
+      require (submittedEntries[newHash][msg.sender][entry] == false); // ... but not this exact entry
+    }
+    // TODO: Require minimum stake, that is (much) more than the cost required to defend the valid submission.
+    // Check the ticket is a winning one.
+    // TODO Figure out how to uncomment the next line, but not break tests sporadically.
```

Comment:
If I forward time by 3600, and this is in place, no submissions will be accepted (because some time will have elapsed between `forwardTime` and the submission being made.

Conversely, if I forward time by e.g. 3500, sometimes the tests will randomly fail because the hash will happen to not be small enough.